File: python/utils/logger.py
from pathlib import Path
from datetime import datetime


class Logger:
    project_root = Path(__file__).resolve().parents[2]
    log_directory = project_root / "data" / "logs_python"
    log_directory.mkdir(parents=True, exist_ok=True)
    file_name = datetime.now().strftime("%Y-%m-%d_%H-%M-%S.log")
    path = log_directory / file_name
    file = open(path, "a", encoding="utf-8")

    @staticmethod
    def _write(level: str, chat_id: int | None, text: str):
        current_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        chat = "-" if chat_id is None else str(chat_id)
        message = f"[{current_time}] [{level}] [{chat}] {text}"

        print(message)
        Logger.file.write(message + "\n")
        Logger.file.flush()

    @staticmethod
    def info(chat_id: int | None, text: str):
        Logger._write("INFO", chat_id, text)

    @staticmethod
    def warning(chat_id: int | None, text: str):
        Logger._write("WARNING", chat_id, text)

    @staticmethod
    def error(chat_id: int | None, text: str):
        Logger._write("ERROR", chat_id, text)

File: cpp/src/logger.cpp
//
//  logger.cpp
//  Kufar Telegram Notifier
//
//  Created by Macintosh on 27.07.2026.
//

#include <chrono>
#include <filesystem>
#include <format>

#include "logger.hpp"

std::string getCurrentTime(){
    auto now = std::chrono::time_point_cast<std::chrono::seconds>(
        std::chrono::system_clock::now()
    );
    return std::format("{:%Y-%m-%d_%H-%M-%S}", now);
}

std::string Logger::path;
std::string Logger::fileName;
std::ofstream Logger::file;

LogStream::LogStream(LogLevel level){
    this->level = level;
}

LogStream::~LogStream(){
    std::string message = buffer.str();
    std::string time = getCurrentTime();
    std::string text = "[" + time + "] [" + Logger::levelToString(level) + "] " + message;
    Logger::write(text, level);
}

void Logger::init(){
    std::filesystem::create_directories("data/logs_cpp/");
    Logger::fileName = getCurrentTime() + ".log";
    Logger::path = "data/logs_cpp/" + Logger::fileName;

    Logger::file.open(Logger::path);
    if (!file.is_open()){
        std::cerr << "Cannot open log file\n";
    }
}

std::string Logger::levelToString(LogLevel level){
    if (level==LogLevel::info) return "INFO";
    if (level==LogLevel::warning) return "WARNING";
    if (level==LogLevel::error) return "ERROR";
    if (level==LogLevel::debug) return "DEBUG";

    return "UNKNOWN";
}

void Logger::write(const std::string& text, LogLevel level){
    if (level == LogLevel::info) std::cout << text << std::endl;
    else std::cerr << text << std::endl;
    if (file.is_open()) file << text << std::endl;
}

LogStream Logger::info(){
    return LogStream(LogLevel::info);
}

LogStream Logger::warning(){
    return LogStream(LogLevel::warning);
}

LogStream Logger::error(){
    return LogStream(LogLevel::error);
}

LogStream Logger::debug(){
    return LogStream(LogLevel::debug);
}


